File: addTwo.java


public class addTwo {
    // https://leetcode.com/problems/add-two-numbers/
    public static void main(String[] args) {
        // ListNode l1 = new ListNode(2);
        // ListNode l12 = new ListNode(4);
        // ListNode l13 = new ListNode(3);
        // l1.next = l12;
        // l12.next = l13;
        // ListNode l2 = new ListNode(5);
        // ListNode l22 = new ListNode(6);
        // ListNode l23 = new ListNode(4);
        // l2.next = l22;
        // l22.next = l23;

        // ListNode l1 = new ListNode(0, null);
        // ListNode l2 = new ListNode(0, null);

        // ListNode l1 = new ListNode(9, null);
        // ListNode l12 = new ListNode(9, null);
        // ListNode l13 = new ListNode(1, null);
        // l1.next = l12;
        // l12.next = l13;
        // ListNode l2 = new ListNode(1, null);

        ListNode l1 = new ListNode(3, null);
        ListNode l12 = new ListNode(7, null);
        ListNode l2 = new ListNode(9);
        ListNode l22 = new ListNode(2);
        l1.next = l12;
        l2.next = l22;
        ListNode result = addTwoNumbers(l1, l2);
        System.out.print("");
    }

    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0, null);
        ListNode trail = new ListNode(0, null);
        int resultint;
        int carry = 0;
        // Need to assign value to result as it is guaranteed to exist
        if ((l1.val + l2.val) > 9) {
            resultint = l1.val + l2.val - 10;
            carry = 1;
        } else {
            resultint = l1.val + l2.val;
        }
        result.val = resultint;
        l1 = l1.next;
        l2 = l2.next;
        // Iterate over the lists when they have same degree of values
        if ((l1 != null) || (l2 != null) || (carry == 1)) {
            // If we have more than one node, attach result to trail
            result.next = trail;
        }
        // Loop over the parallel linked lists
        while ((l1 != null) && (l2 != null)) {
            if ((l1.val + l2.val + carry) > 9) {
                resultint = (l1.val + l2.val + carry) - 10;
                carry = 1;
            } else {
                resultint = l1.val + l2.val + carry;
                carry = 0;
            }
            trail.val = resultint;
            if ((l1.next != null) || (l2.next != null) || (carry == 1)) {
                ListNode trail2 = new ListNode(0, null);
                trail.next = trail2;
                trail = trail.next;
            }
            l1 = l1.next;
            l2 = l2.next;
        }
        if (l1 != null) {
            while (l1 != null) {
                if (l1.val + carry > 9) {
                    resultint = l1.val + carry - 10;
                    carry = 1;
                } else {
                    resultint = l1.val + carry;
                    carry = 0;
                }
                trail.val = resultint;
                if ((l1.next != null) || (carry == 1)) {
                    ListNode trail2 = new ListNode(0, null);
                    trail.next = trail2;
                    trail = trail.next;
                }
                l1 = l1.next;
            }
        }
        if (l2 != null) {
            while (l2 != null) {
                if (l2.val + carry > 9) {
                    resultint = l2.val + carry - 10;
                    carry = 1;
                } else {
                    resultint = l2.val + carry;
                    carry = 0;
                }
                trail.val = resultint;
                if ((l2.next != null) || (carry == 1)) {
                    ListNode trail2 = new ListNode(0, null);
                    trail.next = trail2;
                    trail = trail.next;
                }
                l2 = l2.next;
            }
        }
        if (carry == 1) {
            trail.val = 1;
        }
        return result;
    }
}

File: addtwo.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next



def addTwoNumbers(self, l1, l2):
    carry = int(0)
    result = ListNode(self)
    resultIterator = ListNode(self)
    resultIterator = result
    while (l1.next != None) and (l2.next != None):
        resultIterator.val = l1.val + l2.val + carry
        if carry == 1:
            carry = 0
        if resultIterator.val >= 10:
            carry = 1
            resultIterator.val -= 10
        resNext = ListNode(self)
        resultIterator.next = resNext
        resultIterator = resultIterator.next
        l1 = l1.next
        l2 = l2.next
    # Both the lists still have a val, just no next
    resultIterator.val = l1.val + l2.val + carry
    if carry == 1:
        carry = 0
    if resultIterator.val >= 10:
        carry = 1
        resultIterator.val -= 10
    if (l1.next != None) or (l2.next != None):
        resNext = ListNode(self)
        resultIterator.next = resNext
        resultIterator = resultIterator.next
    elif carry == 1:
        resNext = ListNode(self)
        resNext.val = carry
        resultIterator.next = resNext
    # we've caught the last nodes that are adjacent
    if l1.next != None:
        l1 = l1.next
        while l1.next != None:
            resultIterator.val = l1.val + carry
            if carry == 1:
                carry = 0
            if resultIterator.val >= 10:
                carry = 1
                resultIterator.val -= 10
            resNext = ListNode(self)
            resultIterator.next = resNext
            resultIterator = resultIterator.next
            l1 = l1.next
        # catch last val
        resultIterator.val = l1.val + carry
        if resultIterator.val >= 10:
            resNext = ListNode(self)
            resultIterator.val -= 10
            resNext.val = 1
            resultIterator.next = resNext
    if l2.next != None:
        l2 = l2.next
        while l2.next != None:
            resultIterator.val = l2.val + carry
            if carry == 1:
                carry = 0
            if resultIterator.val >= 10:
                carry = 1
                resultIterator.val -= 10
            resNext = ListNode(self)
            resultIterator.next = resNext
            resultIterator = resultIterator.next
            l2 = l2.next
        resultIterator.val = l2.val + carry
        if resultIterator.val >= 10:
            resNext = ListNode(self)
            resultIterator.val -= 10
            resNext.val = 1
            resultIterator.next = resNext
    return result

def main():
    # l1 = ListNode(2)
    # l12 = ListNode(4)
    # l1.next = l12
    # l13 = ListNode(3)
    # l12.next = l13
    # l2 = ListNode(5)
    # l22 = ListNode(6)
    # l2.next = l22
    # l23 = ListNode(4)
    # l22.next = l23
    l1 = ListNode(9)
    l2 = ListNode(9)
    l22 = ListNode(9)
    l2.next = l22
    l23 = ListNode(9)
    l22.next = l23
    addTwoNumbers(None, l1, l2)

if __name__ == "__main__":
    main()
